patches/api/0421-Paper-Item-API.patch

```diff
@@ -0,0 +1,582 @@
++     */
++    @Nullable
++    Integer duration();
++}
+diff --git a/src/main/java/io/papermc/paper/inventory/item/properties/TropicalFishVariant.java b/src/main/java/io/papermc/paper/inventory/item/properties/TropicalFishVariant.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..3ae8e35cb98c2b39714f3347ad49a474e505b45a
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/inventory/item/properties/TropicalFishVariant.java
+@@ -0,0 +1,29 @@
++package io.papermc.paper.inventory.item.properties;
++
++import org.bukkit.DyeColor;
++import org.bukkit.entity.TropicalFish;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * Represents a serialized tropical fish.
++ */
++public interface TropicalFishVariant {
++
++    /**
++     * Gets the pattern color
++     * @return color
++     */
++    @NotNull DyeColor patternColor();
++
++    /**
++     * Gets the body color
++     * @return color
++     */
++    @NotNull DyeColor bodyColor();
++
++    /**
++     * Gets the type
++     * @return type
++     */
++    @NotNull TropicalFish.Pattern type();
++}
+diff --git a/src/main/java/io/papermc/paper/property/MutablePropertyView.java b/src/main/java/io/papermc/paper/property/MutablePropertyView.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..d73a2276a3e56905e9595f85f5532f06f8dc583d
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/property/MutablePropertyView.java
+@@ -0,0 +1,56 @@
++package io.papermc.paper.property;
++
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++public interface MutablePropertyView extends PropertyView, Cloneable {
++
++    /**
++     * Unsets a property key.
++     *
++     * @param key key
++     */
++    void unset(@NotNull PropertyKey<?> key);
++
++    /**
++     * Sets the value of the given property key.
++     *
++     * @param key   key
++     * @param value value
++     * @param <T>   type
++     */
++    <T> void set(@NotNull PropertyKey<T> key, @Nullable T value);
++
++    /**
++     * Creates a new property key. TODO: I don't like this
++     *
++     * @param key
++     * @param <T>
++     * @return
++     */
++    @NotNull
++    <T> T create(@NotNull PropertyKey<T> key);
++
++    /**
++     * TODO: I don't like this
++     *
++     * @param key
++     * @param <T>
++     * @return
++     */
++    @NotNull
++    <T> T getOrCreate(@NotNull PropertyKey<T> key);
++
++    /**
++     * Clones the given property view.
++     *
++     * @return cloned value
++     */
++    @NotNull
++    MutablePropertyView clone();
++
++    /**
++     * Clears the property view.
++     */
++    void clear();
++}
+diff --git a/src/main/java/io/papermc/paper/property/PropertyKey.java b/src/main/java/io/papermc/paper/property/PropertyKey.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..4d02e565476a04cdccf78d700a14aa45586f7233
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/property/PropertyKey.java
```

Comment:
I'm not sure its a good idea to generalize this much. I understand that both the item and later entity property APIs will be similar, but they shouldn't be interchangeable, like I shouldn't be able to put an item property on an entity and vice-versa. I think they should have specific types that aren't really related at all inheritance-wise.